cmd/ax/upgrade.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"archive/tar"
+	"bufio"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/kardianos/osext"
+)
+
+var (
+	gitOrganisationName = "egnyte"
+	gitRepositoryName   = "ax"
+)
+
+func getLatestReleaseData() *github.RepositoryRelease {
+	client := github.NewClient(nil)
+
+	release, _, err := client.Repositories.GetLatestRelease(context.Background(), gitOrganisationName, gitRepositoryName)
+	if err != nil {
+		fmt.Printf("Couldn't fetch release information: %v", err)
+		os.Exit(1)
+	}
+
+	return release
+}
+
+func getLatestAssetLink() (string, error) {
+	release := getLatestReleaseData()
+
+	for _, element := range release.Assets {
+		assetLink := element.GetBrowserDownloadURL()
+		if strings.Contains(assetLink, runtime.GOOS) && strings.Contains(assetLink, runtime.GOARCH) {
+			return assetLink, nil
+		}
+	}
+
+	return "", errors.New("assset link couldn't be found")
+}
+
+func getLatestReleaseTag() string {
+	release := getLatestReleaseData()
+	return release.GetTagName()
+}
+
+func downloadFile(filePath string, url string) error {
+
+	out, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if _, err = io.Copy(out, resp.Body); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func upgradeVersion() error {
+	latestVersion := getLatestReleaseTag()
+
+	if version == latestVersion {
+		fmt.Println("Latest version is already installed.")
+		return nil
+	}
+
+	fmt.Printf("New version detected. Current version: %s. Latest version: %s\n", version, latestVersion)
+	inputScanner := bufio.NewScanner(os.Stdin)
+	fmt.Print("Do you want to upgrade Ax to the latest version? (yes/no): ")
+	inputScanner.Scan()
+
+	if text := inputScanner.Text(); strings.ToLower(text) != "yes" {
+		fmt.Println("Aborting the upgrade.")
+		os.Exit(1)
+	}
+
+	latestAssetURL, err := getLatestAssetLink()
+	if err != nil {
+		return err
+	}
+
+	extractionPath, err := ioutil.TempDir("/tmp", "egnyte_ax")
+	if err != nil {
+		return err
+	}
+	downloadPath := fmt.Sprintf("%s/ax.tar.gz", extractionPath)
```

Comment:
This could be still improved but the logic in the `downloadFile` has to be improved as well.